### PR TITLE
Routing all AR ledgers to request editor for manual ledger checking

### PR DESCRIPTION
**Pull request by marctemp**
https://eaflood.atlassian.net/browse/SFI-1659

### CODE REVIEW

File: app/processing/requires-manual-ledger-check.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const util = require('util')
+const getCompletedPaymentRequests = require('./get-completed-payment-requests')
+
+const requiresManualLedgerCheck = async (paymentRequest) => {
+  console.log(`Payment requests: ${util.inspect(paymentRequest, false, null, true)}`)
+  let requiresManualCheck = false
+  if (paymentRequest.value <= 0) {
+    requiresManualCheck = true
```
Comment (marctemp) [inline app/processing/requires-manual-ledger-check.js:8]: Remove the `let` and change to `return true`

Early return pattern is advised for Javascript due to its async nature, reducing the requirement for `else` statements plus it helps to differentiate between edge case

File: app/processing/route-to-request-editor.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ const sendDebtMessage = require('../messaging/send-debt-message')
 const holdAndReschedule = require('../reschedule')
```
Comment (marctemp) [inline app/processing/route-to-request-editor.js:4]: This file is now defunct as we have created a better named `route-debt-to-request-editor` file

File: app/processing/route-debt-to-request-editor.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const db = require('../data')
+const { getHoldCategoryId } = require('../holds')
+const sendDebtMessage = require('../messaging/send-debt-message')
+const holdAndReschedule = require('../reschedule')
+const util = require('util')
+
+const routeToRequestEditor = async (paymentRequest) => {
```
Comment (marctemp) [inline app/processing/route-debt-to-request-editor.js:7]: name should be `routeDebtToRequestEditor`

comment left around removing defunct `route-to-request-editor` file so be careful with cleaning up function names in `index.js` and tests

File: app/processing/route-manual-ledger-to-request-editor.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+const holdAndReschedule = require('../reschedule')
+const util = require('util')
+
+const routeManualLedgerToRequestEditor = async (paymentRequest) => {
+  const transaction = await db.sequelize.transaction()
+  try {
+    const manualLedgerMessage = { paymentRequest: paymentRequest.deltaPaymentRequest, paymentRequests: paymentRequest.completedPaymentRequests }
+    await sendManualLedgerMessage(manualLedgerMessage)
+    console.log('Payment request routed to request editor:', util.inspect(manualLedgerMessage, false, null, true))
+    const holdCategoryId = await getHoldCategoryId(paymentRequest.deltaPaymentRequest.schemeId, 'Manual ledger hold', transaction)
```
Comment (marctemp) [inline app/processing/route-manual-ledger-to-request-editor.js:13]: Consider pulling `paymentRequest.deltaPaymentRequest` as a variable above and using throughout

File: app/messaging/update-requests-awaiting-manual-ledger-check.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+const updateRequestsAwaitingManualLedgerCheck = async (paymentRequest) => {
+  const orginalPaymentRequest = paymentRequest.paymentRequest
+
+  const checkPaymentRequest = await db.paymentRequest.findOne({ where: { invoiceNumber: orginalPaymentRequest.invoiceNumber } })
+  if (!checkPaymentRequest) {
+    throw new Error(`No payment request matching invoice number: ${paymentRequest.invoiceNumber}`)
+  }
+
+  const scheduleId = paymentRequest.scheduleId
```
Comment (marctemp) [inline app/messaging/update-requests-awaiting-manual-ledger-check.js:14]: Coul pull these out on 1 line, `const { scheduleId,  paymentRequests  } = paymentRequest`

File: app/processing/process-payment-requests.js [hunk 6]
```diff
@@ -16,13 +19,15 @@ const processPaymentRequest = async (scheduledPaymentRequest) => {
   const paymentRequests = await transformPaymentRequest(scheduledPaymentRequest.paymentRequest)
```
Comment (marctemp) [inline app/processing/process-payment-requests.js:19]: Could pull out `const { deltaPaymentRequest, completedPaymentRequests }  = paymentRequests`

File: app/processing/delta/copy-payment-request.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+const getDefaultLedger = require('./get-default-ledger')
```
Comment (marctemp) [inline app/processing/delta/copy-payment-request.js:1]: See comment in `app/processing/delta/index.js`

File: app/processing/delta/index.js [hunk 8]
```diff
@@ -1,42 +1,35 @@
...
 const zeroValueSplit = require('./zero-value-split')
+const copyPaymentRequest = require('./copy-payment-request')
 
 const calculateDelta = (paymentRequest, previousPaymentRequests) => {
   const invoiceLines = getInvoiceLines(paymentRequest, previousPaymentRequests)
 
   const lineDeltas = calculateLineDeltas(invoiceLines)
   const overallDelta = calculateOverallDelta(invoiceLines)
   const updatedPaymentRequest = copyPaymentRequest(paymentRequest, overallDelta, lineDeltas)
+  const deltaPaymentRequest = JSON.parse(JSON.stringify(updatedPaymentRequest))
```
Comment (marctemp) [inline app/processing/delta/index.js:15]: I think we need a better way of doing this, whether that's creating a reusable function to do so -- used 3 times throughout and likely to increase -- again, not in the scope of this ticket, more technical debt

File: app/processing/transform-payment-request.js [hunk 9]
```diff
@@ -7,11 +7,12 @@ const transformPaymentRequest = async (paymentRequest) => {
   // if yes, need to treat as post payment adjustment and calculate Delta which can result in payment request splitting
   const previousPaymentRequests = await getCompletedPaymentRequests(paymentRequest.schemeId, paymentRequest.frn, paymentRequest.marketingYear, paymentRequest.agreementNumber, paymentRequest.paymentRequestNumber)
   if (previousPaymentRequests.length) {
-    const updatedPaymentRequests = calculateDelta(paymentRequest, previousPaymentRequests)
-    return enrichPaymentRequests(updatedPaymentRequests, previousPaymentRequests)
+    const deltaPaymentRequests = calculateDelta(paymentRequest, previousPaymentRequests)
```
Comment (marctemp) [inline app/processing/transform-payment-request.js:10]: Could pull this out as `const { deltaPaymentRequests, completedPaymentRequests } = calculateDelta(paymentRequest, previousPaymentRequests)` to match what `calculateDelta` is producing then pulling it together again for the return, `return { deltaPaymentRequests, completedPaymentRequests}`

File: test/integration/local/processing/process-payment-requests-manual-ledger.test.js [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+      planned: moment().subtract(1, 'day')
+    }
+  })
+
+  afterAll(async () => {
+    await db.sequelize.truncate({ cascade: true })
+    await db.sequelize.close()
+  })
+
+  test('should process manual ledger request and create hold if useManualLedgerCheck equals true when delta value is < 0', async () => {
```
Comment (marctemp) [inline test/integration/local/processing/process-payment-requests-manual-ledger.test.js:79]: This should be glued to the bottom of `test/integration/local/processing/process-payment-requests.test.js`

File: changelog/db.changelog-1.17.xml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.9.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+  <changeSet author="The Dream Team" id="1">
+    <tagDatabase tag="v1.17.0" />
+  </changeSet>
+  <changeSet author="The Dream Team" id="2">
+    <insert tableName="holdCategories">
+      <column name="holdCategoryId" value="10" />
+      <column name="schemeId" value="2" />
```
Comment (johnwatson484) [inline changelog/db.changelog-1.17.xml:9]: Also needs a copy for scheme id `1`

File: changelog/db.changelog.xml [hunk 12]
```diff
@@ -21,4 +21,5 @@
   <include  file="changelog/db.changelog-1.14.xml"/>
   <include  file="changelog/db.changelog-1.15.xml"/>
   <include  file="changelog/db.changelog-1.16.xml"/>
+  <include  file="changelog/db.changelog-1.17.xml"/>
```
Comment (johnwatson484) [inline changelog/db.changelog.xml:24]: As a heads up, recent merges have already used `1.17` so will need updating

File: changelog/db.changelog-1.17.xml [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.9.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+  <changeSet author="The Dream Team" id="1">
```
Comment (johnwatson484) [inline changelog/db.changelog-1.17.xml:3]: 🥇

File: provision.azure.yaml [hunk 14]
```diff
@@ -5,3 +5,4 @@ resources:
     - name: return
     - name: debt
     - name: qc
+    - name: manual-check
```
[the remark this answers is not in the document]
Comment (johnwatson484) [reply provision.azure.yaml:8]: needs to match the env var in docker compose and helm values.  so `QCMANUALCHECK_TOPIC_ADDRESS`, `qcManualCheckTopicAddress` means we should have `qcManualCheck`

There's also a `manualCheck` entry needed.

**Review by johnwatson484 — CHANGES_REQUESTED** [review]
Added some extra feedback.

Also there are many conflicts with main that need resolving.

File: helm/ffc-pay-processing/values.yaml [hunk 15]
```diff
@@ -49,6 +49,11 @@ container:
   debtTopicAddress: ffc-pay-debt-data
   qcTopicAddress: ffc-pay-quality-check
```
Comment (johnwatson484) [inline helm/ffc-pay-processing/values.yaml:50]: These names are tidied up in a subticket elsewhere but for this PR to work, `qcTopicAddress` needs to be `ffc-pay-debt-data-response`

File: docker-compose.yaml [hunk 16]
```diff
@@ -24,12 +24,16 @@ services:
       DEBT_TOPIC_ADDRESS: ${DEBT_TOPIC_ADDRESS:-ffc-pay-debt-data}${MESSAGE_QUEUE_SUFFIX}
       QC_TOPIC_ADDRESS: ${QC_TOPIC_ADDRESS:-ffc-pay-quality-check}${MESSAGE_QUEUE_SUFFIX}
```
Comment (johnwatson484) [inline docker-compose.yaml:25]: These names are tidied up in a subticket elsewhere but for this PR to work, QC needs to be `ffc-pay-debt-data-response`

File: docker-compose.yaml [hunk 17]
```diff
@@ -24,12 +24,16 @@ services:
...
       QC_SUBSCRIPTION_ADDRESS: ${QC_SUBSCRIPTION_ADDRESS:-ffc-pay-processing}${MESSAGE_QUEUE_SUFFIX}
+      MANUALCHECK_TOPIC_ADDRESS: ${MANUALCHECK_TOPIC_ADDRESS:-ffc-pay-manual-check-data}${MESSAGE_QUEUE_SUFFIX}
+      QCMANUALCHECK_TOPIC_ADDRESS: ${MANUALCHECK_TOPIC_ADDRESS:-ffc-pay-manual-ledger-check}${MESSAGE_QUEUE_SUFFIX}
```
Comment (johnwatson484) [inline docker-compose.yaml:28]: Needs to be `ffc-pay-quality-check`

File: helm/ffc-pay-processing/values.yaml [hunk 18]
```diff
@@ -49,6 +49,11 @@ container:
...
   qcSubscriptionAddress: ffc-pay-processing
+  useManualLedgerCheck: false
+  manualCheckTopicAddress: ffc-pay-manual-check-data
+  qcManualCheckTopicAddress: ffc-pay-manual-ledger-check
```
Comment (johnwatson484) [inline helm/ffc-pay-processing/values.yaml:54]: Needs to be `ffc-pay-quality-check`

File: app/processing/requires-manual-ledger-check.js (hunk 1 above)
Comment (AbidemiAdio) [reply app/processing/requires-manual-ledger-check.js:8]: SonarQube flagged two early return true as code smell, it advised they should be combined as single return.

File: docs/asyncapi.yaml [hunk 19]
```diff
@@ -115,6 +119,212 @@ components:
                 accountCode:
                   type: string
                   description: Unique account code to pay from
+    ManualLedgerCheck:
```
Comment (marctemp) [inline docs/asyncapi.yaml:122]: This has brought up questions over the content of the payloads being sent and how they can be reduced to aid in maintainability -- payload properties which need assessed recorded before

File: docs/asyncapi.yaml [hunk 20]
```diff
@@ -115,6 +119,212 @@ components:
...
+      contentType: application/json
+      payload:
+        type: object
+        properties:
+          deltaPaymentRequest:
```
Comment (marctemp) [inline docs/asyncapi.yaml:127]: Can we strip this right back to only having properties which can/will be changed by automated ledger assignment?

File: docs/asyncapi.yaml [hunk 21]
```diff
@@ -115,6 +119,212 @@ components:
...
+                ledger:
+                  type: string
+                  description: Target ledger for settlement, AP for Accounts Payable and AR for Accounts Receivable
+                marketingYear:
+                  type: number
+                  description: Scheme year for request
+                originalSettlementDate:
+                  type: string
+                  description: Date original invoice settled, optional
+                paymentRequestId:
```
Comment (marctemp) [inline docs/asyncapi.yaml:285]: Id should not be returned as it means nothing and only causes confusion between micro-services when `paymentRequestId: 1` in `ffc-pay-processing` is the corresp. `paymentRequestId: 2` record in `ffc-pay-request-editor`

File: docs/asyncapi.yaml [hunk 22]
```diff
@@ -115,6 +119,212 @@ components:
...
+                recoveryDate:
+                  type: string
+                  description: Date recovery identified, optional
+                sbi:
+                  type: number
+                  description: Single Business Identifier, optional
+                schedule:
+                  type: string
+                  description: Payment frequency instruction for DAX, eg Q4 for quarterly payments
+                scheme:
```
Comment (marctemp) [inline docs/asyncapi.yaml:303]: Does this need returned?

File: docs/asyncapi.yaml [hunk 23]
```diff
@@ -115,6 +119,212 @@ components:
...
+                    active:
+                      type: boolean
+                      description: Whether the scheme is live
+                    name:
+                      type: string
+                      description: Name of the scheme
+                    schemeId:
+                      type: number
+                      description: Unique id for the scheme
+                schemeId:
```
Comment (marctemp) [inline docs/asyncapi.yaml:316]: Dup. property returned in `scheme` object above

File: docs/asyncapi.yaml [hunk 24]
```diff
@@ -115,6 +119,212 @@ components:
...
+                originalSettlementDate:
+                  type: string
+                  description: Date original invoice settled, optional
+                paymentRequestId:
+                  type: number
+                  description: Unique id for request
+                paymentRequestNumber:
+                  type: number
+                  description: Version of payment request starting with 1, anything greater than 1 is an adjustment
+                received:
```
Comment (marctemp) [inline docs/asyncapi.yaml:291]: Type `timestamp` in `completedPaymentRequest` but type `string` in `deltaPaymentRequest`

i.e. `Thu Mar 17 2022 12:14:41 GMT+0000 (Coordinated Universal Time)` vs `'2022-03-17T12:14:41.490Z'`

File: docs/asyncapi.yaml [hunk 25]
```diff
@@ -115,6 +119,212 @@ components:
...
+              originalSettlementDate:
+                type: string
+                description: Date original invoice settled, optional
+              paymentRequestId:
+                type: number
+                description: Unique id for request
+              paymentRequestNumber:
+                type: number
+                description: Version of payment request starting with 1, anything greater than 1 is an adjustment
+              received:
```
Comment (marctemp) [inline docs/asyncapi.yaml:188]: Type timestamp in completedPaymentRequest but type string in deltaPaymentRequest

i.e. Thu Mar 17 2022 12:14:41 GMT+0000 (Coordinated Universal Time) vs '2022-03-17T12:14:41.490Z'

File: docs/asyncapi.yaml (hunk 19 above)
Comment (AbidemiAdio) [reply docs/asyncapi.yaml:122]: https://github.com/DEFRA/ffc-pay-processing/pull/55#discussion_r828300221

Payload constructed as advised by John

**Review by johnwatson484 — APPROVED** [review]

**Pull request merged**